File: 1545/FindKthBitInNthString.cpp
/*
https://leetcode.com/problems/find-kth-bit-in-nth-binary-string/?envType=daily-question&envId=2024-10-19
*/

class Solution {
public:

    string invert(string str){
        string res="";
        for(int i=0; i<str.length(); ++i){
            if(str[i]=='0') res+="1";
            else res+="0";
        }
        return res;
    }

    char findKthBit(int n, int k) {
        vector<string>all;
        string prev="0";
        int count=1;
        all.push_back(prev);
        while(count!=n){
            string invertPrev=invert(prev);
            reverse(invertPrev.begin(), invertPrev.end());
            string next=prev+"1"+invertPrev;
            all.push_back(next);
            prev=next;
            count++;
        }
        return all[n-1][k-1];
    }
};

File: 1/1000s/15s/1545/FindKthBitInNthString.java
/*
https://leetcode.com/problems/find-kth-bit-in-nth-binary-string/?envType=daily-question&envId=2024-10-19
*/

class Solution {

    public static String reverseString(String str){
        StringBuilder reversedString=new StringBuilder();
        int lengthOfStr=str.length();
        for(int i=lengthOfStr-1; i>=0; --i){
            reversedString.append(str.charAt(i));
        }
        return String.valueOf(reversedString);
    }

    public static String invertString(String str){
        StringBuilder resultantInvertedString=new StringBuilder();
        int lengthOfStr=str.length();
        for(int i=0; i<lengthOfStr; ++i){
            if(str.charAt(i)=='1') resultantInvertedString.append("0");
            else resultantInvertedString.append("1");
        }
        return String.valueOf(resultantInvertedString);
    }

    public char findKthBit(int n, int k) {
        ArrayList<String>allStringsUntilN=new ArrayList<>();
        StringBuilder previousString=new StringBuilder("0");
        int count=1;
        allStringsUntilN.add(String.valueOf(previousString));
        while(count!=n){
            String invertedPrev=invertString(String.valueOf(previousString));
            String reversedString=reverseString(invertedPrev);
            StringBuilder nextString=new StringBuilder();
            nextString.append(previousString);
            nextString.append("1");
            nextString.append(reversedString);
            allStringsUntilN.add(String.valueOf(nextString));
            previousString=nextString;
            count++;
        }
        return allStringsUntilN.get(n-1).charAt(k-1);
    }
}